xtask/src/semver_check.rs

```diff
@@ -0,0 +1,294 @@
+use std::{
+    fs, io::Write, path::{Path, PathBuf}
+};
+
+use esp_metadata::Chip;
+use rustdoc_types::ItemEnum;
+
+use crate::{Package, cargo::CargoArgsBuilder};
+
+pub fn generate_baseline(
+    workspace: &PathBuf,
+    packages: Vec<Package>,
+    chips: Vec<Chip>,
+) -> Result<(), anyhow::Error> {
+    for package in packages {
+        log::info!("Generating API baseline for {package}");
+
+        for chip in &chips {
+            log::info!("Chip = {}", chip.to_string());
+            let package_name = package.to_string();
+            let package_path = crate::windows_safe_path(&workspace.join(&package_name));
+
+            let current_path = build_doc_json(package, chip, &package_path)?;
+
+            remove_unstable_items(&current_path)?;
+
+            let file_name = if package.chip_features_matter() {
+                format!("{}", chip.to_string())
+            } else {
+                "api".to_string()
+            };
+
+            let to_path = PathBuf::from(&package_path).join(format!("api-baseline/{}.json.gz", file_name));
+            fs::create_dir_all(to_path.parent().unwrap())?;
+
+            log::debug!("Compress into {current_path:?}");
+            let mut encoder = flate2::write::GzEncoder::new(fs::File::create(to_path)?, flate2::Compression::default());
+            encoder.write_all(&std::fs::read(current_path)?)?;
+
+            if !package.chip_features_matter() {
+                break;
+            }
+        }
+    }
+
+    Ok(())
+}
+
+pub fn check(
+    workspace: &PathBuf,
+    packages: Vec<Package>,
+    chips: Vec<Chip>,
+) -> Result<(), anyhow::Error> {
+    for package in packages {
+        log::info!("Semver-check API for {package}");
+
+        for chip in &chips {
+            log::info!("Chip = {}", chip.to_string());
+            let package_name = package.to_string();
+            let package_path = crate::windows_safe_path(&workspace.join(&package_name));
+
+            let current_path = build_doc_json(package, chip, &package_path)?;
+            remove_unstable_items(&current_path)?;
+
+            let file_name = if package.chip_features_matter() {
+                format!("{}", chip.to_string())
+            } else {
+                "api".to_string()
+            };
+
+            let baseline_path_gz = PathBuf::from(&package_path).join(format!("api-baseline/{}.json.gz", file_name));
+
+            let baseline_path = temp_file::TempFile::new()?;
+            let buffer = Vec::new();
+            let mut decoder = flate2::write::GzDecoder::new(buffer);
+            decoder.write_all(&(fs::read(&baseline_path_gz)?))?;
+            fs::write( baseline_path.path(), decoder.finish()?)?;
+
+            let mut check = cargo_semver_checks::Check::new(
+                cargo_semver_checks::Rustdoc::from_path(current_path),
+            );
+            check.set_baseline(cargo_semver_checks::Rustdoc::from_path(baseline_path.path()));
```

Comment:
Oh no, can we avoid the tempfile here, with `from_registry` maybe?

Reply:
Currently no for two reasons
- `instability`'s inability to produce docs only for the stable API (by default, and it even can't by opting in)
- https://github.com/obi1kenobi/cargo-semver-checks/issues/1068

What is the concern about using the checked-into-git baseline?

Reply:
My concern is about decompressing the .gz every time and writing it into a temporary file. This will not be too frequent, I guess, but still I'd like to avoid wearing down flash unless absolutely necessary.

Reply:
maybe I'm missing something but given how generally the GH-runners start from scratch, install stuff and checkout the repo that is not adding too much on top?